Zeppelin: Add Zeppelin image to Spark example

zmerlynn opened the pull request:
This adds a very basic Zeppelin image that works with the existing
Spark example. As can be seen from the documentation, it has a couple
of warts:
- It requires kubectl port-forward (which is unstable across long
  periods of time, at least for me, on this app, bug incoming).
- I needed to roll my own container (none of the existing containers
  exactly matched needs, or even built anymore against modern Zeppelin
  master, and the rest of the example is Spark 1.5).

The image itself is _huge_. One of the further refinements we need to
look at is how to possibly strip the Maven build for this container
down to just the interpreters we care about, because the deps here
are frankly ridiculous.

This might be a case where, if possible, we might want to open an
upstream request to build things dynamically, then use something like
probably the cut the image down considerably. (This might already be
possible, need to poke at whether you can late-bind interpreters
later.)

This is a bullet on #16517

cc @kubernetes/sig-big-data

zmerlynn committed:
Zeppelin: Add Zeppelin image to Spark example

This adds a very basic Zeppelin image that works with the existing
Spark example. As can be seen from the documentation, it has a couple
of warts:

* It requires kubectl port-forward (which is unstable across long
periods of time, at least for me, on this app, bug incoming). See

* I needed to roll my own container (none of the existing containers
exactly matched needs, or even built anymore against modern Zeppelin
master, and the rest of the example is Spark 1.5).

The image itself is *huge*. One of the further refinements we need to
look at is how to possibly strip the Maven build for this container
down to just the interpreters we care about, because the deps here
are frankly ridiculous.

This might be a case where, if possible, we might want to open an
upstream request to build things dynamically, then use something like
probably the cut the image down considerably. (This might already be
possible, need to poke at whether you can late-bind interpreters
later.)

timothysc commented on examples/spark/README.md:
```diff
@@ -172,32 +172,44 @@ you should now see the workers in the UI as well. *Note:* The UI will have links
 
+**Note:** The Zeppelin pod is large, so it may take a while to pull depending on
+  your network. The size of the Zeppelin pod is something we're working on, see
+  issue **TBD**.
```
can reflect back to this issue and chain out.

timothysc replied:
Has this issue been filed, where we can add a link?

zmerlynn replied:
I was trying to figure out where to file this. I can file it in the main k/k repo, but to your later point, if we're going to break something out, it's actually an issue for a separate repo. I'll go ahead and file it here and we can copy it over.

zmerlynn replied:
Filed #17231

timothysc commented on examples/spark/images/base/Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM java:latest
+FROM java:openjdk-8-jdk
```
What version does upstream recommend, as they typically establish a baseline.

zmerlynn replied:
`Building Spark using Maven requires Maven 3.3.3 or newer and Java 7+`

But I was actually pegging it because I didn't want the Zeppelin build process to go all over the process (in the mythical world where an `openjdk-9-jdk` gets released).

zmerlynn replied:
(And yes, I realize we're not building Spark in this image, I'm just pegging the JDK for the runtime.)

zmerlynn replied:
(And I didn't even push a new image, because `:latest` is actually == `:openjdk-8-jdk` right now.)

mattf replied:
simple scan of the zeppelin docs show java 1.7 - https://zeppelin.incubator.apache.org/docs/install/install.html

maybe that means 1.7 only or 1.7+, pinning on java 8 start makes sense

zmerlynn replied:
Yeah, it seems to be function on JDK8, but I could kick it back if you want. Sorry, I did notice that pre-req as well and presumed it meant 1.7+. Other parts of that page are slightly in need of editing, too.

mattf replied:
i'd say stick w/ the 1.7+ assumption until something goes wrong

timothysc commented on examples/spark/images/zeppelin/Dockerfile:
```diff
@@ -0,0 +1,69 @@
+# build break similar to
+# https://www.mail-archive.com/users@zeppelin.incubator.apache.org/msg01586.html
+
+RUN apt-get update \
```
Ohh man, do we really want to install mvn and devel tooling?  Or poke upstream harder?

zmerlynn replied:
All the Dockerfiles I found for Zeppelin built it. This image takes a long time.

My plan was to use this PR as a foil to poke the upstream (especially on `ZEPPELIN-10`, the multi-port issue).

mattf replied:
binaries for spark 1.4.0 are available, which would be preferable

@rnowling have version of zeppelin do you have running and does it work w/ spark 1.5?

zmerlynn replied:
Where? http://zeppelin-project.org/download.html

```
Current binary release
We will provide binary release pretty soon. Until then, please build your own. It's not that difficult.

For install and configure, check install.
```

The v0.5.5 tag (which I just noticed, and may peg to now - it happened while I was working.) has no binaries on GitHub.

mattf replied:
i guess zeppelin-project.org is out of sync w/ https://zeppelin.incubator.apache.org/download.html

rnowling replied:
I last tested Zeppelin with Spark 1.3 or 1.4.  We haven't been using Zeppelin actively.

zmerlynn replied:
Ah, yeah, there are too many pages.

These are based on the old 0.5 tag, which is from July. If they upload a
0.5.5 tag (which has the multiport issue fixed), I'll see if works at all
with Spark 1.5, but I'm not holding out much hope at cross-compatibility.
On Nov 10, 2015 7:26 AM, "Matthew Farrellee" <notifications@github.com>
wrote:

> In examples/spark/images/zeppelin/Dockerfile
> <https://github.com/kubernetes/kubernetes/pull/17025#discussion_r44418976>
> :
>
> > +
> > +ENV MAVEN_HOME /usr/share/maven
> > +
> > +ENV SPARK_MINOR  1.5
> > +ENV SPARK_PATCH  1
> > +ENV SPARK_VER    ${SPARK_MINOR}.${SPARK_PATCH}
> > +ENV HADOOP_MINOR 2.6
> > +ENV HADOOP_PATCH 1
> > +ENV HADOOP_VER   ${HADOOP_MINOR}.${HADOOP_PATCH}
> > +
> > +# libfontconfig is a workaround for
> > +# https://github.com/karma-runner/karma/issues/1270, which caused a
> > +# build break similar to
> > +# https://www.mail-archive.com/users@zeppelin.incubator.apache.org/msg01586.html
> > +
> > +RUN apt-get update \
>
> i guess zeppelin-project.org is out of sync w/
> https://zeppelin.incubator.apache.org/download.html
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17025/files#r44418976>.
>

mattf replied:
@rnowling is spark 1.5 w/ zeppelin 0.5 something you can test out in O(1hr)? if so, please do.

rnowling replied:
@mattf Zeppelin 0.5 doesn't support Spark 1.5 -- there is no Spark 1.5 profile under spark/pom.xml and (at least in my experience), Zeppelin is very specific about the version.  The Zeppelin Dockerfile will need to be updated to 0.5.5 to support Spark 1.5.  I'm building and testing Z 0.5.5 w/ Spark 1.5 locally now.

rnowling replied:
I have Zeppelin 0.5.5 working with Spark 1.5 -- Spark 1.5.1 is bundled as part of the spark-1.5 profile.

zmerlynn replied:
Yes, this Dockerfile is based on master, which I was testing with as of a few days ago and is practically 0.5.5. As I mentioned in https://github.com/kubernetes/kubernetes/pull/17025/files#r44416358, I can shift it over to 0.5.5 now that there's a tag that makes any sense. Zeppelin needs a better release cadence, though - their previous release was *very* old.

zmerlynn replied:
JFYI, if I'm looking at their `create_release.sh` correctly, they're also building for YARN. It's unclear if the YARN support will play well with Standalone, even when they do finally publish binaries.

timothysc commented on examples/spark/images/zeppelin/Dockerfile:
```diff
@@ -0,0 +1,69 @@
+  rm -rf /root/.npm && \
+  echo "Successfully built Zeppelin"
+
+ADD zeppelin-log4j.properties /opt/zeppelin/conf/log4j.properties
```
Same comments as other entry around config.  I would almost like to hold off further PRs until that is in place.

zmerlynn replied:
Yeah, I may take a trip sideways while we figure out how to get a more polished story around a lot of things in this space. I wanted to get a vertical slice up first.

timothysc commented:
So, I'm starting to debate if the upstream k8's repo makes the most sense for these?  Or whether we should find a new home and converse with the broader community.  IMHO, the helm conversation also helps to further this topic.

zmerlynn commented:
@timothysc: Are you basically talking about #17027? I opened it right after I sent off this PR. :)

zmerlynn commented:
Please hold. I just learned that contrary to the documentation, `8081` is no longer in use. (Maybe? I swear I saw traffic.)

zmerlynn commented:
Ah, egg/face. Our service proxy is broken for this case, which confused this issue. I can `port-forward 8080:8080` alone and it works fine, but the proxy isn't handling it.

zmerlynn commented:
PTAL. Now green, and updated for just `8080`. #17032 is blocking me from switching to the application proxy immediately, but this is still simpler.

timothysc commented:
I'd like to get some eyeballs on this that use it all the time to verify and offer their opinion.  @rnowling, @willb

zmerlynn commented:
PTAL, updated image to peg to v0.5.5. Now that it's a constant release image, I could just maven build the release myself and dump *that* in GCS, but you'd have to trust me. :)

timstclair commented:
It looks like other Tim St. Clair has this one - thanks!

willb commented:
I'm able to fire up the notebook interface, but am seeing network errors between the notebook and the interpreter service when I try and execute any code.

zmerlynn commented:
@willb: Can you be more specific, or cut and paste? What workload are you trying, what's your configuration, etc?

zmerlynn commented:
And just to level-set: I'm not stating this is working for all workloads yet. There are almost certainly bugs. It took considerable effort just to get it working for the one case.

willb commented:
@zmerlynn sure, I'm running Kubernetes master from a checkout (with this patch applied) on Fedora 22, using `hack/local-up-cluster.sh`.  I'm not really doing anything interesting with Zeppelin (even just executing a basic line of Scala code, like `val five = 5` doesn't work).  Here's the Thrift exception I'm seeing:

```

org.apache.thrift.transport.TTransportException 
  at org.apache.thrift.transport.TIOStreamTransport.read(TIOStreamTransport.java:132) 
  at org.apache.thrift.transport.TTransport.readAll(TTransport.java:86) 
  at org.apache.thrift.protocol.TBinaryProtocol.readAll(TBinaryProtocol.java:429) 
  at org.apache.thrift.protocol.TBinaryProtocol.readI32(TBinaryProtocol.java:318) 
  at org.apache.thrift.protocol.TBinaryProtocol.readMessageBegin(TBinaryProtocol.java:219) 
  at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:69) 
  at org.apache.zeppelin.interpreter.thrift.RemoteInterpreterService$Client.recv_interpret(RemoteInterpreterService.java:220) 
  at org.apache.zeppelin.interpreter.thrift.RemoteInterpreterService$Client.interpret(RemoteInterpreterService.java:205) 
  at org.apache.zeppelin.interpreter.remote.RemoteInterpreter.interpret(RemoteInterpreter.java:211) 
  at org.apache.zeppelin.interpreter.LazyOpenInterpreter.interpret(LazyOpenInterpreter.java:93) 
  at org.apache.zeppelin.notebook.Paragraph.jobRun(Paragraph.java:207) 
  at org.apache.zeppelin.scheduler.Job.run(Job.java:170) 
  at org.apache.zeppelin.scheduler.RemoteScheduler$JobRunner.run(RemoteScheduler.java:304) 
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) 
  at java.util.concurrent.FutureTask.run(FutureTask.java:266) 
  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) 
  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) 
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) 
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) 
  at java.lang.Thread.run(Thread.java:745) 
```

We've seen interpreter-notebook communication problems with Zeppelin on bare metal in the past, so it might be a Zeppelin issue now (i.e., not necessarily an issue with the networking configuration for the example); I just wanted to let you know as soon as possible that this fails a basic smoke test for me.  What kinds of workloads are you testing with?

zmerlynn commented:
The included README in this PR has the notebook I used. I'm on mobile for a bit or I'd paste it in. I don't actually have a lot of experience running `local-up`, either. You may want to run through the whole Spark example and see if it's functional for local, because my suspicion is something is broken there. (Does it start DNS? This absolutely requires DNS to be functional, and most real apps will.)

zmerlynn commented:
No, `local-up-cluster.sh` doesn't start DNS, see:
http://kubernetes.io/v1.0/docs/getting-started-guides/locally.html

willb commented:
Thanks, @zmerlynn; I'll find another place to test.

binarybana commented on examples/spark/images/zeppelin/Dockerfile:
```diff
@@ -0,0 +1,66 @@
+ENV HADOOP_PATCH  1
+ENV HADOOP_VER    ${HADOOP_MINOR}.${HADOOP_PATCH}
+
+RUN curl -fsSL http://archive.apache.org/dist/maven/maven-3/${MAVEN_VERSION}/binaries/apache-maven-${MAVEN_VERSION}-bin.tar.gz | tar xzf - -C /usr/share \
```
You use curl here, though you install it further down on line 46. (though I changed the base image, so perhaps this isn't a problem for you on `java`, but then you should probably not try and install it again on line 46.

zmerlynn replied:
`curl` is installed on `buildpack-deps`, so this part is a noop. Fixed, though.

binarybana commented:
FWIW, this pod and service are working well so far for us with some modifications to match our needs:
- rebase this onto another base image rather than `java`
- Bumped to Spark 1.5.2 (which necessitated using an un-merged PR of Zeppelin)
- Using some additional Hadoop jars for AWS `s3a://` support

Thanks!

binarybana commented:
Although I just remembered that the `kubectl port-forward` didn't work as written in the docs:
```
$ kubectl port-forward zeppelin-controller-brwmj 8080:8080
I1111 12:39:51.356549   18547 portforward.go:225] Forwarding from 127.0.0.1:8080 -> 8080
I1111 12:39:51.356725   18547 portforward.go:225] Forwarding from [::1]:8080 -> 8080
I1111 12:39:58.740803   18547 portforward.go:251] Handling connection for 8080
I1111 12:39:58.742427   18547 portforward.go:251] Handling connection for 8080
I1111 12:39:58.742507   18547 portforward.go:251] Handling connection for 8080
I1111 12:39:58.742527   18547 portforward.go:251] Handling connection for 8080
I1111 12:39:58.742579   18547 portforward.go:251] Handling connection for 8080
I1111 12:39:58.742578   18547 portforward.go:251] Handling connection for 8080
E1111 12:39:58.849566   18547 portforward.go:310] An error occurred forwarding 8080 -> 8080: error forwarding port 8080 to pod zeppelin-controller-brwmj_default, uid : unable to do port forwarding: socat not found.
I1111 12:40:00.049220   18547 portforward.go:251] Handling connection for 8080
E1111 12:40:30.049760   18547 portforward.go:262] Error creating error stream for port 8080 -> 8080: Timeout occured
```
But this may be because we switched the docker base image to `ubuntu:15.04` from `java`. But I just checked and socat is not in `java:openjdk-8-jdk` either?

zmerlynn commented:
@binarybana: Awesome!

> rebase this onto another base image rather than java

Why? I'm mostly asking because I'm trying to create a turnkey solution (I'm next going to package this as a helm package and a DM template), so I'm curious what you had to change.

> Bumped to Spark 1.5.2 (which necessitated using an un-merged PR of Zeppelin)

Yeah, I'm not going to bump this to 1.5.2 until this is merged, and I'd actually rather not do that until Zeppelin ships another release. Given their existing cadence, that might be a while. Did you have a particular 1.5.2 issue/feature you needed?

> Using some additional Hadoop jars for AWS s3a:// support

Yeah, this will be easier once we have #831 and #6477.

zmerlynn commented:
@binarybana: That's actually a complaint that `socat` isn't found on the node the pod was scheduled on (it's a complaint from `kubelet`). Is this a roll-your-own configuration? All of our Salt configurations have `socat` as a required package dependency.

binarybana commented:
@zmerlynn: We wanted to use a platform (ubuntu) that the bioinformatics people around here were more comfortable with, and it's not easy to see what base image java is built off of (though I suspect Debian of some version). The amount of changes necessary were minimal (just curl really), which was really nice. I'll have to check out helm at some point.

> Yeah, I'm not going to bump this to 1.5.2 until this is merged, and I'd actually rather not do that until Zeppelin ships another release. Given their existing cadence, that might be a while. Did you have a particular 1.5.2 issue/feature you needed?

There are some data integrity issues fixed in 1.5.2, and upgrade is apparently _[strongly recommended](https://spark.apache.org/releases/spark-release-1-5-2.html) (to my eyes [this](https://issues.apache.org/jira/browse/SPARK-11153) was pretty important). 

> That's actually a complaint that socat isn't found on the node the pod was scheduled on (it's a complaint from kubelet). Is this a roll-your-own configuration? All of our Salt configurations have socat as a required package dependency.

This is a heavily modified and updated CoreOS cloudformation template based off of the Kubernetes [unsupported AWS script](https://github.com/kubernetes/kubernetes/blob/release-1.0/docs/getting-started-guides/aws/cloudformation-template.json). I'm not sure how to get socat installed on CoreOS, though my plan is to probably just expose Zeppelin through a `NodePort` anyways.

zmerlynn commented:
@binarybana: Thanks for the info. The And yes, the `java` image is actually Debian Jessie underneath.

binarybana commented:
Just in case this might help anyone like myself. I also ran into some trouble with attempting to connect to the Spark-cluster from outside the cluster (for example through a `NodePort` from an interactive `pyspark --master 'spark://<kubernetes_master_ip>:30001'` session on your workstation).

You will see errors like this show up from Spark-Master: 
```
15/11/12 22:10:45 ERROR ErrorMonitor: dropping message [class akka.actor.ActorSelectionMessage] for non-local recipient [Actor[akka.tcp://sparkMaster@spark-master:30002/]] arriving at [akka.tcp://sparkMaster@spark-master:30002] inbound addresses are [akka.tcp://sparkMaster@spark-master:7077]
akka.event.Logging$Error$NoCause$
15/11/12 22:11:06 WARN ReliableDeliverySupervisor: Association with remote system [akka.tcp://sparkDriver@172.16.6.100:57177] has failed, address is now gated for [5000] ms. Reason: [Disassociated] 
```

The problem is that the Akka network backend that Spark uses depends on connections coming to host `spark-master` and port `7077` and will break otherwise. This was mentioned by @zmerlynn in #16517 , but I had to spend an hour or two banging into it before I could figure it out for myself.

I'm not sure if there's any good workaround for this use case, except modifying `/etc/hosts` on the external client to point `spark-master` to `localhost` and using a `kubectl port-forward <spark-master-pod> 7077:7077`. 

This isn't directly relevant to this PR, but it might be nice to mention this in the docs to avoid people trying to do this blindly.

zmerlynn commented:
@timothysc: Ping on a review, thanks. (It would be nice to iterate.)

timothysc commented:
I'll review 1st thing in the a.m.

timothysc commented on examples/spark/README.md:
```diff
@@ -172,32 +172,44 @@ you should now see the workers in the UI as well. *Note:* The UI will have links
+a web notebook frontend or the traditional Spark command line. Read more about
+[Zeppelin](https://zeppelin.incubator.apache.org/) and the
+[Spark architecture](https://spark.apache.org/docs/latest/cluster-overview.html)
+to understand more.
```
s/Read more about X and the Y to understand more./See X and Y for more details.

zmerlynn replied:
Thanks. I knew that sentence had gotten away from me. :)

timothysc commented on examples/spark/README.md:
```diff
@@ -217,6 +229,35 @@ SparkContext available as sc, HiveContext available as sqlContext.
+**Also Note**: On GKE, `kubectl port-forward` may not be stable over long
+  periods of time. If you see Zeppelin go into `Disconnected` state (there will
+  be a red dot on the top right as well), the `port-forward` probably
+  failed and needs to be restarted. See #12179.
```
Might want to call this "known issues"

zmerlynn replied:
I'll break out a section.

timothysc commented:
minor nits then lgtm.  

I do think we should start solving the distribution problem before we expand our scope to more BD projects.

zmerlynn commented:
@timothysc: PTAL, I think I addressed the nits.

@binarybana: Thanks for the comments, BTW. I really appreciate someone in the wild using it. :) I incorporated a comment in the README that notes that the `spark-master` service is only usable inside the cluster. This is a common pattern, and was intentional, FWIW - for most cluster services, we try to show examples that don't expose the service to the world unless we're using an external load balancer, and if there was an ELB, I would need auth on it.

timothysc commented:
lgtm.

zmerlynn commented:
@k8s-bot test this

zmerlynn commented:
@k8s-bot test this

zmerlynn commented:
Manually verified `examples` test. Adding e2e-not-required, no idea what's going with PR builder.
